{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6671ab6c-83b9-4163-9d73-03be6c54b410",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Project: Ascites\n",
    "# content: Data analysis\n",
    "\n",
    "# author: kraesing\n",
    "# mail: lau.kraesing.vestergaard@regionh.dk\n",
    "# GitHub: https://github.com/kraesing\n",
    "\n",
    "##########################################################################"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "931a9b48-cacb-482f-9b12-d654ebd54b09",
   "metadata": {},
   "outputs": [],
   "source": [
    "### IMPORT LIBRARIES\n",
    "\n",
    "import warnings\n",
    "warnings.filterwarnings(\"ignore\")\n",
    "\n",
    "# Libraries for file management.\n",
    "import os\n",
    "import glob \n",
    "import fnmatch\n",
    "import shutil \n",
    "\n",
    "# Libraries for data management.\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "from numpy.random import seed\n",
    "\n",
    "# Libraries for statistical analysis.\n",
    "import scipy.stats as stats\n",
    "import scipy.spatial.distance import squareform, pdist\n",
    "\n",
    "# Libraries for plotting.\n",
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "import matplotlib.lines as mlines\n",
    "\n",
    "# Setting params for plotting\n",
    "sns.set_theme(style=\"whitegrid\")\n",
    "sns.set_context(\"notebook\")\n",
    "plt.rcParams[\"font.family\"] = \"Helvetica\"\n",
    "sns.set_palette(\"colorblind\")\n",
    "sns.set_color_codes(palette=\"muted\")\n",
    "\n",
    "# Libraries for machine learning.\n",
    "import tensorflow as tf\n",
    "\n",
    "# Libraries for Natural Language Processing (NLP)\n",
    "from gensim.models import Doc2Vec\n",
    "from gensim.models.doc2vec import TaggedDocument\n",
    "from nltk import word_tokenize\n",
    "from sklearn.feature_extraction.text import CountVectorizer\n",
    "from sklearn.feature_extraction.text import TfidfVectorizer\n",
    "from sklearn.metrics.pairwise import cosine_similarity\n",
    "\n",
    "# Libraries for Network analysis\n",
    "from pyvis.network import Network\n",
    "from itertools import product\n",
    "import networkx as nx\n",
    "\n",
    "# Co-mutational plotting\n",
    "from comut import comut\n",
    "\n",
    "# Conformation\n",
    "print(\"Libraries imported!\")\n",
    "\n",
    "##########################################################################"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3e21a7ae-8517-4ef4-819b-ca7020de4c52",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Setting seed \n",
    "\n",
    "np.random.seed(34)\n",
    "tf.random.set_seed(34)\n",
    "##########################################################################"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "76990897-58cb-49b7-afc4-e122dea4e5e4",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Change directory\n",
    "\n",
    "os.chdir(path + \"IonTorrent5.18\")\n",
    "\n",
    "##########################################################################"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1da6bf60-687b-4fd2-a95a-5beedad335b3",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Loading data\n",
    "\n",
    "os.chdir(os.getcwd() + path_to_files)\n",
    "\n",
    "samples_ascites_ = glob.glob(\"*600005*\")\n",
    "samples_tissue_ = glob.glob(\"*600006*\")\n",
    "samples_plasma_ = glob.glob(\"*600004*\")\n",
    "\n",
    "##########################################################################"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "187c1efa-8a4a-4ee8-b3e1-69c88a7280ad",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Loading clinical data\n",
    "\n",
    "df_patients_id = pd.read_excel(path_to_clinical_data, sheet_name=\"OvCA\", dtype=\"str\")\n",
    "list_patients_id = list(set(df_patients_id[\"ID\"]))\n",
    "\n",
    "# Remove nan element\n",
    "removed_element = list_patients_id.pop(0)\n",
    "\n",
    "# Introduce leading zeros for ID with the length of 1.\n",
    "list_patients_id = [\"0\"+x if len(x) < 2 else x for x in list_patients_id]\n",
    "\n",
    "# Creating empty lists\n",
    "samples_tissue = []\n",
    "samples_ascites = []\n",
    "samples_plasma = []\n",
    "\n",
    "for i in samples_tissue_:\n",
    "    if i[19:21] in list_patients_id:\n",
    "        samples_tissue.append(i)\n",
    "        \n",
    "for i in samples_ascites_:\n",
    "    if i[19:21] in list_patients_id:\n",
    "        samples_ascites.append(i)\n",
    "        \n",
    "for i in samples_plasma_:\n",
    "    if i[19:21] in list_patients_id:\n",
    "        samples_plasma.append(i)        \n",
    "\n",
    "##########################################################################"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e7c16356-ffec-41fa-8a53-159935bedf6d",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Loading Histological data\n",
    "\n",
    "histology = pd.read_excel(r\"histologi_.xlsx\", sheet_name=\"Sheet4\")\n",
    "histology[\"ID\"] = [\"0\"+str(x) if len(str(x)) == 1 else str(x)\n",
    "                   for x in histology[\"ID\"]]\n",
    "\n",
    "##########################################################################"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "971f716c-5420-4384-89c7-607dd3b71187",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Counting initial variants\n",
    "\n",
    "counting_variants(samples_ascites, samples_tissue)\n",
    "\n",
    "##########################################################################"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f692a415-13bb-434f-86e2-6ad37b3e436d",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Running variant filtering for ascites samples\n",
    "\n",
    "interval_start = 0\n",
    "interval_stop = 21\n",
    "\n",
    "mutation_list_ascites = []\n",
    "mutation_rest_ascites = []\n",
    "resc_ascites = []\n",
    "merged_ascites = []\n",
    "rescued_ascites = []\n",
    "ascites_samples = []\n",
    "\n",
    "for sample in samples_ascites:\n",
    "    print(\"Analyzing sample: {}\".format(sample[interval_start:interval_stop]))\n",
    "    ascites_samples.append(sample[interval_start:interval_stop])\n",
    "    df = pd.read_csv(sample, sep=\"\\t\")\n",
    "    df[\"samplename\"] = sample[interval_start:interval_stop]\n",
    "    df1 = Initial_cleaning(df)\n",
    "    df1 = Gene_cleaning(df1)\n",
    "    df1 = Original_filtering(df=df1, allele_freq=(7), coverage=(200))\n",
    "    # df1 = df1.loc[~df1[\"Clinical_significance\"].isin(list_benigns)]\n",
    "    df_pass = df1.loc[df1[\"true_variant\"] == \"PASS\"]\n",
    "    if len(df1) != 0:\n",
    "        df2 = Variant_rescue_cleaning(df1)\n",
    "        apply_fishers_test_(df=df2, strandbias_threshold=0.05)\n",
    "        mutation_rest_ascites.append(df2)\n",
    "\n",
    "    mutation_list_ascites.append(df_pass)\n",
    "\n",
    "for i in mutation_rest_ascites:\n",
    "    _ = i.copy()\n",
    "    _ = _.loc[~(_[\"Var+\"] == \"0\")]\n",
    "    _ = _.loc[~(_[\"Var-\"] == \"0\")]\n",
    "    p_value = _.loc[_[\"true_variant\"] == \"Above p-value\"]\n",
    "    low_phred = _.loc[_[\"true_variant\"] == \"Low Phred Score\"]\n",
    "    base_coverage = _.loc[_[\"true_variant\"] == \"Low base coverage\"]\n",
    "    below_threshold = _.loc[_[\"true_variant\"] ==\n",
    "                            \"Below threshold for allele frequency\"]\n",
    "    low_overall = _.loc[_[\"true_variant\"] ==\n",
    "                        \"Low overall coverage (caution warrant)\"]\n",
    "\n",
    "    if len(low_phred) != 0:\n",
    "        Variant_rescue_filtering(df=low_phred, Frequency=2.5, Pvalue=0.0001,\n",
    "                                 PhredScore=40, Coverage=750, suffix=\"Low Phred score\")\n",
    "\n",
    "    if len(base_coverage) != 0:\n",
    "        Variant_rescue_filtering(df=base_coverage, Frequency=2.5, Pvalue=0.0001,\n",
    "                                 PhredScore=200, Coverage=70, suffix=\"Low base-coverage\")\n",
    "\n",
    "    if len(low_overall) != 0:\n",
    "        Variant_rescue_filtering(df=low_overall, Frequency=2.5, Pvalue=0.0001, PhredScore=200,\n",
    "                                 Coverage=70, suffix=\"Low base-coverage overall (caution warrant)\")\n",
    "\n",
    "    if len(p_value) != 0:\n",
    "        Variant_rescue_filtering(df=p_value, Frequency=0.075, Pvalue=0.05,\n",
    "                                 PhredScore=12.5, Coverage=500, suffix=\"Low p-value\")\n",
    "\n",
    "    if len(below_threshold) != 0:\n",
    "        Variant_rescue_filtering(df=below_threshold, Frequency=2.5, Pvalue=0.0001,\n",
    "                                 PhredScore=400, Coverage=1000, suffix=\"Below allele_threshold\")\n",
    "\n",
    "    frames = [p_value, low_phred, base_coverage, below_threshold, low_overall]\n",
    "    merged_ascites_rest = pd.concat(frames)\n",
    "    merged_ascites.append(merged_ascites_rest.loc[merged_ascites_rest[\"Rescue_filtering\"].str.contains(\"PASS Rescue filtering\")])\n",
    "\n",
    "##########################################################################"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ad5cae1b-e574-4ef9-bd05-825e45039ecd",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Running variant filtering for tumor tissue samples\n",
    "\n",
    "mutation_list_tissue = []\n",
    "mutation_rest_tissue = []\n",
    "resc_tissue = []\n",
    "merged_tissue = []\n",
    "rescued_tissue = []\n",
    "tissue_samples = []\n",
    "\n",
    "for sample in samples_tissue:\n",
    "    print(\"Analyzing sample: {}\".format(sample[interval_start:interval_stop]))\n",
    "    tissue_samples.append(sample[interval_start:interval_stop])\n",
    "    df = pd.read_csv(sample, sep=\"\\t\")\n",
    "    df[\"samplename\"] = sample[interval_start:interval_stop]\n",
    "    df = remove_artifactual_variants(df)\n",
    "    df = df.dropna(subset=[\"Ref+/Ref-/Var+/Var-\"])\n",
    "    df1 = Initial_cleaning(df)\n",
    "    df1 = Gene_cleaning(df1)\n",
    "    df1 = Original_filtering(df=df1, allele_freq=(11), coverage=(200))\n",
    "    # df1 = df1.loc[~df1[\"Clinical_significance\"].isin(list_benigns)]\n",
    "    df_pass = df1.loc[df1[\"true_variant\"] == \"PASS\"]\n",
    "    if len(df1) != 0:\n",
    "        df2 = Variant_rescue_cleaning(df1)\n",
    "        apply_fishers_test_(df=df2, strandbias_threshold=0.05)\n",
    "        mutation_rest_tissue.append(df2)\n",
    "\n",
    "    mutation_list_tissue.append(df_pass)\n",
    "\n",
    "\n",
    "for i in mutation_rest_tissue:\n",
    "    _ = i.copy()\n",
    "    _ = _.loc[~(_[\"Var+\"] == \"0\")]\n",
    "    _ = _.loc[~(_[\"Var-\"] == \"0\")]\n",
    "    p_value_t = _.loc[_[\"true_variant\"] == \"Above p-value\"]\n",
    "    low_phred_t = _.loc[_[\"true_variant\"] == \"Low Phred Score\"]\n",
    "    base_coverage_t = _.loc[_[\"true_variant\"] == \"Low base coverage\"]\n",
    "    below_threshold_t = _.loc[_[\"true_variant\"]\n",
    "                              == \"Below threshold for allele frequency\"]\n",
    "    low_overall_t = _.loc[_[\"true_variant\"] ==\n",
    "                          \"Low overall coverage (caution warrant)\"]\n",
    "\n",
    "    if len(low_phred_t) != 0:\n",
    "        Variant_rescue_filtering(df=low_phred_t, Frequency=2.5, Pvalue=0.0001,\n",
    "                                 PhredScore=40, Coverage=750, suffix=\"Low Phred score\")\n",
    "\n",
    "    if len(base_coverage_t) != 0:\n",
    "        Variant_rescue_filtering(df=base_coverage_t, Frequency=2.5, Pvalue=0.0001,\n",
    "                                 PhredScore=200, Coverage=70, suffix=\"Low base-coverage\")\n",
    "\n",
    "    if len(low_overall_t) != 0:\n",
    "        Variant_rescue_filtering(df=low_overall_t, Frequency=2.5, Pvalue=0.0001, PhredScore=200,\n",
    "                                 Coverage=70, suffix=\"Low base-coverage overall (caution warrant)\")\n",
    "\n",
    "    if len(p_value_t) != 0:\n",
    "        Variant_rescue_filtering(df=p_value_t, Frequency=0.075, Pvalue=0.05,\n",
    "                                 PhredScore=12.5, Coverage=500, suffix=\"Low p-value\")\n",
    "\n",
    "    if len(below_threshold_t) != 0:\n",
    "        Variant_rescue_filtering(df=below_threshold_t, Frequency=2.5, Pvalue=0.0001,\n",
    "                                 PhredScore=400, Coverage=1000, suffix=\"Below allele_threshold\")\n",
    "\n",
    "    if len(low_phred_t + base_coverage_t + p_value_t + below_threshold_t + low_overall_t) == 0:\n",
    "        low_phred_t[\"Rescue_filtering\"] = \"Empty\"\n",
    "\n",
    "    frames_t = [p_value_t, low_phred_t, base_coverage_t,\n",
    "                below_threshold_t, low_overall_t]\n",
    "    merged_tissue_rest = pd.concat(frames_t)\n",
    "    merged_tissue.append(merged_tissue_rest.loc[merged_tissue_rest[\"Rescue_filtering\"].str.contains(\"PASS Rescue filtering\")])\n",
    "\n",
    "##########################################################################"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "64da0031-2467-4adb-9101-42efddaa3128",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Running variant filtering for plasma samples. NOT INCLUDED IN THE STUDY.\n",
    "\n",
    "mutation_list_plasma = []\n",
    "mutation_rest_plasma = []\n",
    "resc_plasma = []\n",
    "merged_plasma = []\n",
    "rescued_plasma = []\n",
    "plasma_samples = []\n",
    "\n",
    "for sample in samples_plasma:\n",
    "    print(\"Analyzing sample: {}\".format(sample[interval_start:interval_stop]))\n",
    "    plasma_samples.append(sample[interval_start:interval_stop])\n",
    "    df = pd.read_csv(sample, sep=\"\\t\")\n",
    "    df[\"samplename\"] = sample[interval_start:interval_stop]\n",
    "    df = remove_artifactual_variants(df)\n",
    "    df = df.dropna(subset=[\"Ref+/Ref-/Var+/Var-\"])\n",
    "    df1 = Initial_cleaning(df)\n",
    "    df1 = Gene_cleaning(df1)\n",
    "    df1 = Original_filtering(df=df1, allele_freq=(7), coverage=(200))\n",
    "    # df1 = df1.loc[~df1[\"Clinical_significance\"].isin(list_benigns)]\n",
    "    df_pass = df1.loc[df1[\"true_variant\"] == \"PASS\"]\n",
    "    if len(df1) != 0:\n",
    "        df2 = Variant_rescue_cleaning(df1)\n",
    "        apply_fishers_test_(df=df2, strandbias_threshold=0.05)\n",
    "        mutation_rest_plasma.append(df2)\n",
    "\n",
    "    mutation_list_plasma.append(df_pass)\n",
    "\n",
    "for i in mutation_rest_plasma:\n",
    "    _ = i.copy()\n",
    "    _ = _.loc[~(_[\"Var+\"] == \"0\")]\n",
    "    _ = _.loc[~(_[\"Var-\"] == \"0\")]\n",
    "    p_value_p = _.loc[_[\"true_variant\"] == \"Above p-value\"]\n",
    "    low_phred_p = _.loc[_[\"true_variant\"] == \"Low Phred Score\"]\n",
    "    base_coverage_p = _.loc[_[\"true_variant\"] == \"Low base coverage\"]\n",
    "    below_threshold_p = _.loc[_[\"true_variant\"]\n",
    "                              == \"Below threshold for allele frequency\"]\n",
    "    low_overall_p = _.loc[_[\"true_variant\"] ==\n",
    "                          \"Low overall coverage (caution warrant)\"]\n",
    "\n",
    "    if len(low_phred_p) != 0:\n",
    "        Variant_rescue_filtering(df=low_phred_p, Frequency=2.5, Pvalue=0.0001,\n",
    "                                 PhredScore=40, Coverage=750, suffix=\"Low Phred score\")\n",
    "\n",
    "    if len(base_coverage_p) != 0:\n",
    "        Variant_rescue_filtering(df=base_coverage_p, Frequency=2.5, Pvalue=0.0001,\n",
    "                                 PhredScore=200, Coverage=70, suffix=\"Low base-coverage\")\n",
    "\n",
    "    if len(low_overall_p) != 0:\n",
    "        Variant_rescue_filtering(df=low_overall_p, Frequency=2.5, Pvalue=0.0001, PhredScore=200,\n",
    "                                 Coverage=70, suffix=\"Low base-coverage overall (caution warrant)\")\n",
    "\n",
    "    if len(p_value_p) != 0:\n",
    "        Variant_rescue_filtering(df=p_value_p, Frequency=0.075, Pvalue=0.05,\n",
    "                                 PhredScore=12.5, Coverage=500, suffix=\"Low p-value\")\n",
    "\n",
    "    if len(below_threshold_p) != 0:\n",
    "        Variant_rescue_filtering(df=below_threshold_p, Frequency=2.5, Pvalue=0.0001,\n",
    "                                 PhredScore=400, Coverage=1000, suffix=\"Below allele_threshold\")\n",
    "\n",
    "    if len(low_phred_p + base_coverage_p + p_value_p + below_threshold_p + low_overall_p) == 0:\n",
    "        low_phred_p[\"Rescue_filtering\"] = \"Empty\"\n",
    "\n",
    "    frames_p = [p_value_p, low_phred_p, base_coverage_p,\n",
    "                below_threshold_p, low_overall_p]\n",
    "    merged_plasma_rest = pd.concat(frames_p)\n",
    "    merged_plasma.append(merged_plasma_rest.loc[merged_plasma_rest[\"Rescue_filtering\"].str.contains(\"PASS Rescue filtering\")])\n",
    "\n",
    "###############################################################################"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8ba7a366-2e16-4cd3-aec7-8ab37f202440",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Merging variants lists\n",
    "\n",
    "\n",
    "t_ = pd.concat(mutation_list_tissue)\n",
    "a_ = pd.concat(mutation_list_ascites)\n",
    "p_ = pd.concat(mutation_list_plasma)\n",
    "df_combined = pd.concat([t_, a_, p_])\n",
    "\n",
    "t__ = pd.concat(merged_tissue)\n",
    "a__ = pd.concat(merged_ascites)\n",
    "p__ = pd.concat(merged_plasma)\n",
    "df_combined_res = pd.concat([t__, a__, p__])\n",
    "\n",
    "###############################################################################"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "903c0103-b762-456b-90ed-fc4af0936a4f",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Creating a list of columns to be included for subsetting\n",
    "\n",
    "subset_cols = [\"samplename\", \"Locus\", \"Genes1\", \"Genotype\", \"Amino Acid Change\", \"Raw Coverage\", \"Location\",\n",
    "               \"Phred QUAL Score\", \"P-Value\", \"Allele Frequency %\", \"Rescue_filtering\", \"Variant Effect\", \"Clinical_significance\"]\n",
    "\n",
    "###############################################################################"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "68cf19bd-18fb-49be-a6a2-f8a85de685ac",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Subsetting the data\n",
    "\n",
    "# Creating a new column, to identify which filter the variant originates from\n",
    "df_combined[\"Rescue_filtering\"] = \"Original\"\n",
    "\n",
    "# Subsetting the data with the columns defined in subset_cols\n",
    "df_combined = df_combined[subset_cols]\n",
    "df_combined_res = df_combined_res[subset_cols]\n",
    "\n",
    "final_data = pd.concat([df_combined, df_combined_res])\n",
    "final_data[\"ID_samplename\"] = final_data[\"samplename\"].apply(lambda x: x[:-3])\n",
    "\n",
    "###############################################################################"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "de277298-0685-4234-802e-feb0aa7a7941",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Checking the number of samples\n",
    "\n",
    "print(final_data[\"ID_samplename\"].value_counts())\n",
    "\n",
    "###############################################################################"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5513ffad-bd39-41d1-9378-040c453d4664",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Inspecting the number of unique genes\n",
    "\n",
    "print(final_data.Genes1.nunique())\n",
    "\n",
    "###############################################################################"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f48bc482-e8c1-4754-8f40-7da7f201e9a0",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Preparing data for matrix formation\n",
    "\n",
    "# Making an ID column (mutation) for each mutation identified\n",
    "final_data[\"mutation\"] = final_data[\"Genes1\"] + \"_\" + \\\n",
    "    final_data[\"Locus\"] + \"_\" + final_data[\"Amino Acid Change\"]\n",
    "final_data[\"values\"] = 1.0\n",
    "\n",
    "# https://stackoverflow.com/questions/51881503/assign-a-dictionary-value-to-a-dataframe-column-based-on-dictionary-key\n",
    "# Creating a dictionary to map a specific value to the dataframe\n",
    "dict_biopsytype = {\"600005\": \"Ascites\",\n",
    "                   \"600006\": \"Tissue\",\n",
    "                   \"600004\": \"Plasma\"}\n",
    "\n",
    "# Assigning new columns\n",
    "final_data[\"sample\"] = final_data.samplename.apply(lambda x: x[-9:-3])\n",
    "final_data[\"samplename_alias\"] = final_data[\"sample\"].map(dict_biopsytype)\n",
    "final_data[\"sample_ID\"] = final_data[\"samplename\"].apply(lambda x: x[-3:])\n",
    "final_data[\"sample_ID\"] = [x.replace(\".\", \"_\")\n",
    "                           for x in final_data[\"sample_ID\"]]\n",
    "final_data[\"samplename_alias\"] = final_data[\"samplename_alias\"] + \\\n",
    "    final_data[\"sample_ID\"]\n",
    "final_data[\"biopsy_type_\"] = final_data[\"sample\"].map(dict_biopsytype)\n",
    "\n",
    "###############################################################################"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c43ed5cd-3042-4e18-bd57-ca4c6fac1146",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Filtering the data to only contain variants from ascites and tumor tissue. \n",
    "\n",
    "final_data = final_data.loc[(final_data[\"samplename_alias\"].str.contains(\n",
    "    \"Ascites\")) | (final_data[\"samplename_alias\"].str.contains(\"Tissue\"))]\n",
    "\n",
    "###############################################################################"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d685b7a0-b321-46b9-b017-a05d58531fb4",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Identifying samples with no variants identified\n",
    "\n",
    "# ONLY FOR ASCITES AND TUMOR\n",
    "combined_sample_list = set(ascites_samples + tissue_samples)\n",
    "\n",
    "# The unique samples found after the analysis.\n",
    "identified_sample_list = set(final_data.samplename.unique())\n",
    "\n",
    "# The following will identify samples that didn't have any variants called in the filtering.\n",
    "set_diff = combined_sample_list.symmetric_difference(identified_sample_list)\n",
    "\n",
    "print(set_diff)\n",
    "\n",
    "###############################################################################"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "02eaccaf-7591-49ab-9ea8-09130a6531ec",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Plotting Venn Diagram for all variants identified from ascites and tumor tissue\n",
    "\n",
    "# FOR ALL VARIANTS\n",
    "# Creating a special column for a Venn diagram\n",
    "final_data[\"for_venn\"] = final_data[\"mutation\"] + final_data[\"sample_ID\"]\n",
    "\n",
    "# Storing the mutations from each tissue type into a set.\n",
    "ascites_venn = set(\n",
    "    final_data.loc[final_data[\"sample\"] == \"600005\"][\"for_venn\"])\n",
    "tissue_venn = set(final_data.loc[final_data[\"sample\"] == \"600006\"][\"for_venn\"])\n",
    "\n",
    "total = len(ascites_venn.union(tissue_venn))\n",
    "\n",
    "# Creating the Venn diagram to show overall overlap of identified variants.\n",
    "plt.figure()\n",
    "venn2([ascites_venn, tissue_venn],\n",
    "      set_labels=(\"Ascites\", \"Tumor\"),\n",
    "      set_colors=(\"#a74e71\", \"#4ea784\"),\n",
    "      subset_label_formatter=lambda x: str(\n",
    "          x) + \"\\n(\" + f\"{(x/total):1.00%}\" + \")\",\n",
    "      alpha=1)\n",
    "\n",
    "# Saving the plot\n",
    "# save_fig(plt, \"Venn_diagram_updated.svg\", transparent_background=True)\n",
    "\n",
    "###############################################################################"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "89367537-cd5a-484c-b94e-1e5b247af49b",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Creating list to filter for pathogenic variants\n",
    "\n",
    "list_variants_1 = [\"Pathogenic\", \"Likely pathogenic\", \"Pathogenic/Likely pathogenic\", \"Pathogeic/Likely pathogenic other\", \"Likely pathogenic other\",\n",
    "                   \"Pathogenic drug response other\", \"Conflicting interpretations of pathogenicity CONF Likely pathogenic Uncertain significance\",\n",
    "                   \"drug response\", \"Conflicting interpretations of pathogenicity CONF Pathogenic Likely pathogenic Uncertain significance\",\n",
    "                   \"Conflicting interpretations of pathogenicity CONF Pathogenic Uncertain significance\", \"Pathogenic/Likely pathogenic other\", \"Likely pathogenic*\"]\n",
    "\n",
    "###############################################################################"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "066d2f03-fc92-4ce6-80ae-b6c397304488",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Filtering the data frame \n",
    "\n",
    "final_data_pat = final_data.loc[final_data[\"Clinical_significance\"].isin(\n",
    "    list_variants_1)]\n",
    "\n",
    "###############################################################################"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bf0ce615-67e4-40db-9f4d-6e38ae3468e5",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Identifying samples with no of pathogenic and likely pathogenic variants\n",
    "\n",
    "# ONLY FOR ASCITES AND TUMOR\n",
    "combined_sample_list = set(ascites_samples + tissue_samples)\n",
    "# The unique samples found after the analysis. \n",
    "identified_sample_list_pat = set(final_data_pat.samplename.unique())\n",
    "\n",
    "# The following will identify samples that didn't have any pathogenic variants in the filtering.\n",
    "set_diff_ = combined_sample_list.symmetric_difference(identified_sample_list_pat)\n",
    "\n",
    "###############################################################################"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "26cf014c-36b2-4a2b-b0b0-1fb5419a751d",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Plotting Venn Diagram for pathogenic variants identified from ascites and tumor tissue\n",
    "\n",
    "# Storing the mutations from each tissue type into a set.\n",
    "ascites_venn_pat = set(\n",
    "    final_data_pat.loc[final_data_pat[\"sample\"] == \"600005\"][\"for_venn\"])\n",
    "tissue_venn_pat = set(\n",
    "    final_data_pat.loc[final_data_pat[\"sample\"] == \"600006\"][\"for_venn\"])\n",
    "\n",
    "total_pat = len(ascites_venn_pat.union(tissue_venn_pat))\n",
    "# Creating a venn diagram to show overall overlap of identified mutations.\n",
    "plt.figure()\n",
    "vd_pat = venn2([ascites_venn_pat, tissue_venn_pat],\n",
    "               set_labels=(\"Ascites\", \"Tumor\"),\n",
    "               set_colors=(\"#a74e71\", \"#4ea784\"),\n",
    "               subset_label_formatter=lambda x: str(\n",
    "                   x) + \"\\n(\" + f\"{(x/total_pat):1.00%}\" + \")\",\n",
    "               alpha=1)\n",
    "\n",
    "# Saving the plot\n",
    "# save_fig(plt, \"Venn_diagram_pathogenic_updated.svg\", transparent_background=True)\n",
    "\n",
    "###############################################################################"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9888133c-d4c6-42ef-8e12-3809003ec024",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Calculating Pearson correlation for all variants and pathogenic variants, respectively. \n",
    "\n",
    "# all variants\n",
    "r, p, df_pearson = pearson_coeff(final_data, \"Ascites\", \"Tissue\")\n",
    "\n",
    "# pathogenic variants\n",
    "r_pat, p_pat, df_pearson_pat = pearson_coeff(final_data_pat, \"Ascites\", \"Tissue\")\n",
    "\n",
    "###############################################################################"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e464e8db-1a23-4959-9b93-dfaef9ec3d49",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Plotting box plot for allele frequency comparison of pathogenic variants between ascites and tumor tissue \n",
    "\n",
    "box_plot_data = final_data_pat[[\"Location\", \"Allele Frequency %\", \"biopsy_type_\", \"for_venn\"]]\n",
    "\n",
    "list_diff_box = set(box_plot_data.loc[box_plot_data[\"biopsy_type_\"] == \"Ascites\"][\"for_venn\"]).symmetric_difference(set(box_plot_data.loc[box_plot_data[\"biopsy_type_\"] == \"Tissue\"][\"for_venn\"]))\n",
    "\n",
    "final_data_pat_box = final_data_pat.loc[final_data_pat[\"for_venn\"].isin(list_diff_box)]\n",
    "\n",
    "palette = {\"Ascites\": \"#a74e71\",\n",
    "           \"Tissue\": \"#4ea784\",\n",
    "           \"Plasma\": \"#a7844e\"}\n",
    "\n",
    "\n",
    "plt.figure(figsize=(2, 10))\n",
    "sns.boxplot(x=\"biopsy_type_\", y=\"Allele Frequency %\",\n",
    "            data=final_data_pat, palette=palette)\n",
    "sns.swarmplot(x=\"biopsy_type_\", y=\"Allele Frequency %\", data=final_data_pat,\n",
    "              palette=palette, edgecolor=\"black\", linewidth=1, marker=\"s\")\n",
    "sns.swarmplot(x=\"biopsy_type_\", y=\"Allele Frequency %\", data=final_data_pat_box,\n",
    "              color=\"black\", edgecolor=\"black\", linewidth=1, marker=\"s\")\n",
    "\n",
    "\n",
    "plt.ylabel(\"Allele frequency (%)\")\n",
    "plt.xlabel(\"\")\n",
    "\n",
    "# Saving the plot\n",
    "# save_fig(plt, \"Boxplot_swarmplot_pat_updated.svg\", False)\n",
    "\n",
    "###############################################################################"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "336e5f20-a5c3-4ccb-baab-97f3a183914e",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Identifying data distribution for parametric or non-parametric test\n",
    "\n",
    "# If the p-value is less than 0.05, the null hypothesis is rejected of the shapiro-Wilk test.\n",
    "# Meaning that the samples does not come from a normal distribution.\n",
    "\n",
    "# SHAPIRO-WILK\n",
    "print(stats.shapiro(\n",
    "    final_data_pat.loc[final_data_pat[\"biopsy_type_\"] == \"Ascites\"][\"Allele Frequency %\"]))\n",
    "\n",
    "print(stats.shapiro(\n",
    "    final_data_pat.loc[final_data_pat[\"biopsy_type_\"] == \"Tissue\"][\"Allele Frequency %\"]))\n",
    "\n",
    "# Mann-Whitney U-Test\n",
    "# https://www.reneshbedre.com/blog/mann-whitney-u-test.html?utm_content=cmp-true\n",
    "\n",
    "s_MW, p_MW = stats.mannwhitneyu(final_data_pat.loc[final_data_pat[\"biopsy_type_\"] == \"Ascites\"][\"Allele Frequency %\"],\n",
    "                                final_data_pat.loc[final_data_pat[\"biopsy_type_\"]\n",
    "                                                   == \"Tissue\"][\"Allele Frequency %\"],\n",
    "                                alternative='two-sided')\n",
    "\n",
    "###############################################################################"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "96f62903-86aa-4642-94e9-4b64b104a4c0",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Plotting Pearson correlation for all identified variants\n",
    "\n",
    "# All variants\n",
    "sns.set_theme(style=\"whitegrid\")\n",
    "sns.lmplot(x=\"AF_Tissue\",\n",
    "           y=\"AF_Ascites\",\n",
    "           data=df_pearson,\n",
    "           markers=\"s\",\n",
    "           line_kws=({\"color\": \"#ffa012\"}),  \n",
    "           scatter_kws=({\"color\": \"#5c5954\"}))\n",
    "\n",
    "sns.scatterplot(x=\"AF_Tissue\", y=\"AF_Ascites\", data=df_pearson,\n",
    "                marker=\"s\", edgecolor=\"black\", color=\"#be6563\")\n",
    "plt.xlabel(\"Tumor -\\n Allele frequency (%)\", fontsize=12)\n",
    "plt.ylabel(\"Ascites -\\n Allele frequency (%)\", fontsize=12)\n",
    "plt.xlim(-2.5, 102.5)\n",
    "plt.ylim(-2.5, 102.5)\n",
    "plt.legend([\"pearsonr = %s \\n p = %f\" % (r, p)],\n",
    "           fontsize=12, bbox_to_anchor=(1.5, 0.99))\n",
    "\n",
    "sns.scatterplot(x=\"AF_Tissue\", y=\"AF_Ascites\", data=df_pearson_pat,\n",
    "                marker=\"s\", edgecolor=\"black\", color=\"white\")\n",
    "plt.xlabel(\"Tumor Tissue -\\n Allele frequency (%)\", fontsize=12)\n",
    "plt.ylabel(\"Ascites -\\n Allele frequency (%)\", fontsize=12)\n",
    "plt.xlim(-2.5, 102.5)\n",
    "plt.ylim(-2.5, 102.5)\n",
    "\n",
    "# Saving the plot\n",
    "# save_fig(plt, \"correlation_updated.svg\", False)\n",
    "\n",
    "###############################################################################"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f6fb6d44-9ea4-4bef-801c-2c3209056dd3",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Plotting Pearson correlation for pathogenic variants\n",
    "\n",
    "sns.set_theme(style=\"whitegrid\")\n",
    "sns.lmplot(x=\"AF_Tissue\",\n",
    "           y=\"AF_Ascites\",\n",
    "           data=df_pearson_pat,\n",
    "           markers=\"s\",\n",
    "           line_kws=({\"color\": \"#ffa012\"}),\n",
    "           scatter_kws=({\"color\": \"#5c5954\"}))\n",
    "\n",
    "sns.scatterplot(x=\"AF_Tissue\", y=\"AF_Ascites\", data=df_pearson_pat,\n",
    "                marker=\"s\", edgecolor=\"black\", color=\"white\")\n",
    "plt.xlabel(\"Tumor Tissue -\\n Allele frequency (%)\", fontsize=12)\n",
    "plt.ylabel(\"Ascites -\\n Allele frequency (%)\", fontsize=12)\n",
    "plt.xlim(-2.5, 102.5)\n",
    "plt.ylim(-2.5, 102.5)\n",
    "plt.legend([\"pearsonr = %s \\n p = %f\" % (r_pat, p_pat)],\n",
    "           fontsize=12, bbox_to_anchor=(1.5, 0.99))\n",
    "\n",
    "# Saving the plot\n",
    "# save_fig(plt, \"correlation_pathogenic_updated.svg\", False)\n",
    "\n",
    "###############################################################################"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bb254db7-ca0e-4338-9a12-cab0d7cbb8c0",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Creating matrix for all variants from ascites and tumor tissue\n",
    "\n",
    "# Creating pivot table for all variants\n",
    "pivot = pd.pivot_table(\n",
    "    columns=\"samplename_alias\", index=\"mutation\", values=\"values\", data=final_data)\n",
    "\n",
    "# Creating the matrix for samples in absence of variants\n",
    "print(set_diff)\n",
    "set_diff_df = pd.DataFrame(index=set_diff, columns=pivot.columns)\n",
    "set_diff_df[\"samplename\"] = list(set_diff)\n",
    "set_diff_df.index = [x[-9:-3] for x in set_diff_df.index]\n",
    "set_diff_df.index = set_diff_df.index.map(dict_biopsytype)\n",
    "set_diff_df[\"samplename\"] = set_diff_df[\"samplename\"].apply(lambda x: x[-3:])\n",
    "set_diff_df[\"samplename\"] = [x.replace(\".\", \"_\")\n",
    "                             for x in set_diff_df[\"samplename\"]]\n",
    "set_diff_df.index = set_diff_df.index + set_diff_df[\"samplename\"]\n",
    "set_diff_df.drop(columns=[\"samplename\"], inplace=True)\n",
    "\n",
    "# Merging the set_diff with the matrix for samples with variants.\n",
    "pivot = pd.concat([pivot, set_diff_df])\n",
    "\n",
    "# Zero padding the NaN values.\n",
    "pivot.fillna(0, inplace=True)\n",
    "\n",
    "pivot = pivot.rename_axis(None, axis=1)\n",
    "\n",
    "order_ascites = list(pivot.columns.unique()[0:32])\n",
    "order_tissue = list(pivot.columns.unique()[32:])\n",
    "ordered_ = []\n",
    "\n",
    "for i,k in zip(order_ascites, order_tissue):\n",
    "    ordered_.append(i)\n",
    "    ordered_.append(k)\n",
    "\n",
    "pivot = pivot[ordered_]\n",
    "pivot.columns = [col.replace(\"_\", \" \") for col in pivot.columns]\n",
    "pivot.columns = [col.replace(\"Tissue \", \"TT\") for col in pivot.columns]\n",
    "pivot.columns = [col.replace(\"Ascites \", \"A\") for col in pivot.columns]\n",
    "\n",
    "###############################################################################"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "76a1e655-391c-4f20-8fbd-c9073ae6a175",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Creating matrix for pathogenic variants identified from ascites and tumor tissue\n",
    "\n",
    "# Creating pivot table for pathogenic variants \n",
    "pivot_pat = pd.pivot_table(\n",
    "    columns=\"samplename_alias\", index=\"mutation\", values=\"values\", data=final_data_pat)\n",
    "\n",
    " # Zeropadding the NaN values.\n",
    "pivot_pat.fillna(0, inplace=True)\n",
    "\n",
    "# Creating the matrix for samples with the absent of variants\n",
    "print(set_diff_)\n",
    "set_diff_df_pat = pd.DataFrame(index=set_diff_, columns=pivot_pat.index)\n",
    "set_diff_df_pat[\"samplename\"] = list(set_diff_)\n",
    "set_diff_df_pat.index = [x[-9:-3] for x in set_diff_df_pat.index]\n",
    "set_diff_df_pat.index = set_diff_df_pat.index.map(dict_biopsytype)\n",
    "set_diff_df_pat[\"samplename\"] = set_diff_df_pat[\"samplename\"].apply(lambda x: x[-3:])\n",
    "set_diff_df_pat[\"samplename\"] = [x.replace(\".\", \"_\") for x in set_diff_df_pat[\"samplename\"]]\n",
    "set_diff_df_pat.index = set_diff_df_pat.index + set_diff_df_pat[\"samplename\"]\n",
    "set_diff_df_pat.drop(columns=[\"samplename\"], inplace=True)\n",
    "set_diff_df_pat[\"wildtype\"] = 0.1\n",
    "set_diff_df_pat = set_diff_df_pat.T\n",
    "\n",
    "# Merging the set_diff with the matrix for samples with variants.\n",
    "pivot_pat = pd.concat([pivot_pat, set_diff_df_pat], axis=1)\n",
    "pivot_pat = pivot_pat[ordered_]\n",
    "\n",
    "# Zero padding the NaN values.\n",
    "pivot_pat.fillna(0, inplace=True)\n",
    "\n",
    "pivot_pat = pivot_pat.rename_axis(None, axis=1)\n",
    "\n",
    "\n",
    "pivot_pat.columns = [col.replace(\"_\", \" \") for col in pivot_pat.columns]\n",
    "pivot_pat.columns = [col.replace(\"Tissue \", \"TT\") for col in pivot_pat.columns]\n",
    "pivot_pat.columns = [col.replace(\"Ascites \", \"A\") for col in pivot_pat.columns]\n",
    "\n",
    "###############################################################################"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "176c24bd-0578-41aa-8a10-8a2bf1ace61c",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Computing correlation matrix\n",
    "\n",
    "# For all variants\n",
    "corr_mat = pivot.corr().stack().reset_index(name=\"correlation\")\n",
    "corr_mat[\"size\"] = np.nan\n",
    "corr_mat[\"size\"] = [0 if x <= 0 else x for x in corr_mat[\"correlation\"]]\n",
    "corr_mat[\"size\"] = [0.2 if x > 0 and x <= 0.3 else x for x in corr_mat[\"size\"]]\n",
    "corr_mat[\"size\"] = [0.4 if x > 0.3 and x <= 0.5 else x for x in corr_mat[\"size\"]]\n",
    "corr_mat[\"size\"] = [0.6 if x > 0.5 and x <= 0.7 else x for x in corr_mat[\"size\"]]\n",
    "corr_mat[\"size\"] = [0.8 if x > 0.7 and x < 0.9 else x for x in corr_mat[\"size\"]]\n",
    "corr_mat[\"size\"] = [1 if x > 0.9 else x for x in corr_mat[\"size\"]]\n",
    "\n",
    "\n",
    "# For pathogenic variants\n",
    "corr_mat_pat = pivot_pat.corr().stack().reset_index(name=\"correlation\")\n",
    "corr_mat_pat[\"size\"] = np.nan\n",
    "corr_mat_pat[\"size\"] = [0 if x <= 0 else x for x in corr_mat_pat[\"correlation\"]]\n",
    "corr_mat_pat[\"size\"] = [0.2 if x > 0 and x <= 0.3 else x for x in corr_mat_pat[\"size\"]]\n",
    "corr_mat_pat[\"size\"] = [0.4 if x > 0.3 and x <= 0.5 else x for x in corr_mat_pat[\"size\"]]\n",
    "corr_mat_pat[\"size\"] = [0.6 if x > 0.5 and x <= 0.7 else x for x in corr_mat_pat[\"size\"]]\n",
    "corr_mat_pat[\"size\"] = [0.8 if x > 0.7 and x < 0.9 else x for x in corr_mat_pat[\"size\"]]\n",
    "corr_mat_pat[\"size\"] = [1 if x > 0.9 else x for x in corr_mat_pat[\"size\"]]\n",
    "\n",
    "###############################################################################"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bc7e5731-794b-4e07-8835-4b4d48334569",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Plotting correlation\n",
    "\n",
    "scatter_heatplot(data=corr_mat, x= \"level_0\", y=\"level_1\", hue=\"correlation\", title=\"Corr_plot_vlag_.svg\")\n",
    "\n",
    "scatter_heatplot(data=corr_mat_pat, x= \"level_0\", y=\"level_1\", hue=\"correlation\", title=\"Corr_plot_pat_vlag_.svg\")\n",
    "\n",
    "###############################################################################"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "34cc3327-2168-4dec-b60c-d6fccea34753",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Preparing for NLP of all variants\n",
    "\n",
    "# Preparing data for NLP analysis\n",
    "df_nlp = prep_nlp(final_data, set_diff)\n",
    "\n",
    "corpus = []\n",
    "\n",
    "for i in df_nlp[\"Mutation\"]:\n",
    "    corpus.append(i)\n",
    "\n",
    "# Saving variant files for samples separately \n",
    "#saving_mut_files(df_nlp, \"_all_updated\")\n",
    "\n",
    "###############################################################################"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d8b46d82-71f6-42da-99f9-d213813150dc",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Creating object for NLP for all variants\n",
    "\n",
    "# Counting the number of variants identified in each sample.\n",
    "df_nlp[\"Mutation\"].apply(lambda x: len(str(x).split(\", \"))).sum()\n",
    "\n",
    "# Calculating Cosine similarities.\n",
    "co_sim1_df, doc1_df = Cosine_similarity(corpus=corpus)\n",
    "\n",
    "# Getting the ID of the samples as index and column names. \n",
    "id_names = df_nlp.Samplename\n",
    "id_names = id_names.rename(\"id_names\")\n",
    "col_names = df_nlp.Samplename\n",
    "col_names = col_names.rename(\"col_names\")\n",
    "\n",
    "# Renameing the index and the columns.\n",
    "co_sim1_df.index = id_names\n",
    "co_sim1_df.columns = col_names\n",
    "\n",
    "co_sim1_df.columns = [col.replace(\"T\", \"TT\") for col in co_sim1_df.columns]\n",
    "co_sim1_df.index = [idx.replace(\"T\", \"TT\") for idx in co_sim1_df.index]\n",
    "\n",
    "# Plotting the clustermap\n",
    "plt.figure()\n",
    "sns.clustermap(co_sim1_df, cmap=\"Reds\", yticklabels=True, xticklabels=True, annot=False, figsize=(12,12), dendrogram_ratio=0.15)\n",
    "\n",
    "# Saving the plot\n",
    "# save_fig(plt, \"clustermap_all.svg\", False)\n",
    "\n",
    "###############################################################################"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1b6d1807-e3e5-43c1-932a-0a2a26ee8f61",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Preparing for network analysis\n",
    "  \n",
    "# applying the product method\n",
    "print(list(product(co_sim1_df.index, co_sim1_df.columns)))\n",
    "\n",
    "# Creating series for input for the network analysis\n",
    "sources, targets, weights = source_targets_weights(co_sim1_df)\n",
    "\n",
    "###############################################################################"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0f23f341-486e-439a-962e-1709144585b7",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Creating and plotting the network with cosine similarity as weights for all variants. \n",
    "\n",
    "import random\n",
    "import numpy as np\n",
    "\n",
    "# Seed must be run in the same cell for reproducibility\n",
    "seed = 124\n",
    "random.seed(seed)\n",
    "np.random.seed(seed)\n",
    "\n",
    "# Plotting the network\n",
    "plot_network(sources = sources, targets = targets, weights = weights)\n",
    "\n",
    "# Saving the plot\n",
    "#save_fig(plt, \"network_updated.svg\", True)\n",
    "\n",
    "###############################################################################"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ac7f3baa-9dc0-4d2c-b432-de6c847af358",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Preparing for NLP of pathogenic variants\n",
    "\n",
    "# Preparing data for NLP analysis\n",
    "df_nlp_pat = prep_nlp(final_data_pat, set_diff_)\n",
    "df_nlp_pat[\"Mutation\"]\n",
    "corpus_pat = []\n",
    "\n",
    "for i in df_nlp_pat[\"Mutation\"]:\n",
    "    corpus_pat.append(i)\n",
    "\n",
    "# Saving variant files for samples separately \n",
    "saving_mut_files(df_nlp, \"_pat_updated\")\n",
    "\n",
    "###############################################################################"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "051bead6-7843-429e-8c2d-8613de845f71",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Creating object for NLP for pathogenic variants \n",
    "\n",
    "# Counting the number of pathogenic variants that are identified in each sample.\n",
    "df_nlp_pat[\"Mutation\"].apply(lambda x: len(str(x).split(\", \"))).sum()\n",
    "\n",
    "# Calculating Cosine similarities.\n",
    "co_sim1_df_pat, co_sim1_doc = Cosine_similarity(corpus=corpus_pat)\n",
    "\n",
    "# Getting the ID of the samples as index and column names. \n",
    "id_names = df_nlp_pat.Samplename\n",
    "id_names = id_names.rename(\"id_names\")\n",
    "col_names = df_nlp_pat.Samplename\n",
    "col_names = col_names.rename(\"col_names\")\n",
    "\n",
    "# Renameing the index and the columns.\n",
    "co_sim1_df_pat.index = id_names\n",
    "co_sim1_df_pat.columns = col_names\n",
    "\n",
    "co_sim1_df_pat.columns = [col.replace(\"T\", \"TT\") for col in co_sim1_df_pat.columns]\n",
    "co_sim1_df_pat.index = [idx.replace(\"T\", \"TT\") for idx in co_sim1_df_pat.index]\n",
    "\n",
    "# Plotting the clustermap\n",
    "plt.figure()\n",
    "sns.clustermap(co_sim1_df_pat, cmap=\"Reds\", yticklabels=True, xticklabels=True, annot=False, figsize=(12,12), dendrogram_ratio=0.15)\n",
    "\n",
    "# Saving the plot\n",
    "#save_fig(plt, \"clustermap_pat.svg\", False)\n",
    "\n",
    "###############################################################################"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "58576ddc-cce2-4031-8f0a-cba45f034b78",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Preparing for network analysis for pathogenic variants\n",
    "  \n",
    "# applying the product method\n",
    "print(list(product(co_sim1_df_pat.index, co_sim1_df_pat.columns)))\n",
    "\n",
    "# Creating series for input for the network analysis\n",
    "sources_pat, targets_pat, weights_pat = source_targets_weights(co_sim1_df_pat)\n",
    "\n",
    "###############################################################################"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6ee53469-f9c8-4504-9cfe-e4839d49b599",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Creating and plotting the network with cosine similarity as weights for pathogenic variants. \n",
    "\n",
    "# Seed must be run in the same cell for reproducibility\n",
    "seed = 124\n",
    "random.seed(seed)\n",
    "np.random.seed(seed)\n",
    "\n",
    "# Plotting the network\n",
    "plot_network(sources = sources_pat, targets = targets_pat, weights = weights_pat)\n",
    "\n",
    "# Saving the plot\n",
    "#save_fig(plt, \"network_pat_updated.svg\", False)\n",
    "\n",
    "###############################################################################"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ffecabcd-664e-4d8f-a47d-ba79eeaab738",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Preparing data for co-mutational plotting for pathogenic variants\n",
    "\n",
    "# Grouping data based on their sample ID. \n",
    "o_final_data = final_data_pat.copy()\n",
    "\n",
    "add_final_data = pd.DataFrame(index=set_diff_, columns=o_final_data.columns)\n",
    "add_final_data[\"samplename\"] = list(set_diff_)\n",
    "add_final_data[\"samplename_alias\"] = [\"Tissue\"+ x[-3:] if \"600006\" in x else \"Ascites\"+ x[-3:] for x in add_final_data[\"samplename\"]]\n",
    "add_final_data[\"sample_ID\"] = add_final_data[\"samplename\"].apply(lambda x: x[-3:])\n",
    "final_data_pat[\"sample_ID\"] = final_data_pat[\"sample_ID\"].apply(lambda x: x.replace(\"_\", \"\"))\n",
    "add_final_data[\"biopsy_type_\"] = add_final_data[\"samplename_alias\"].apply(lambda x: x[:-3])\n",
    "add_final_data.reset_index(drop=True, inplace=True)\n",
    "add_final_data.fillna(\"[]\", inplace=True)\n",
    "# add_final_data[\"sample_ID\"] = [x.replace(\"_\", \"\") for x in add_final_data.sample_ID]\n",
    "# add_final_data[\"sample_ID_\"] = add_final_data.biopsy_type_.apply(lambda x: x[:1])\n",
    "# add_final_data[\"sample_ID\"] = add_final_data[\"sample_ID_\"] + add_final_data[\"sample_ID\"]\n",
    "\n",
    "# Concatenating data\n",
    "o_final_data = pd.concat([o_final_data, add_final_data])\n",
    "\n",
    "clinical_data_ = pd.read_excel(r\"H:\\PhD\\Work_Packages\\Work_package1_Ascites\\Kliniske_data\\DGCD_data_sampling.xlsx\")\n",
    "o_final_data[\"samplename\"] = o_final_data.samplename.apply(lambda x: x.replace(\".\", \"_\"))\n",
    "o_final_data[\"sample_ID\"] = o_final_data.sample_ID.apply(lambda x: x.replace(\"_\", \"\")).astype(\"int64\")\n",
    "\n",
    "o_final_data = pd.merge(o_final_data, clinical_data_, left_on=\"sample_ID\", right_on=\"ID\", how=\"inner\")\n",
    "o_final_data = o_final_data.replace(\"[]\", np.nan)\n",
    "\n",
    "###############################################################################"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "32737cd3-90d9-47ff-a93d-e6bf8833d07b",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Comutational plotting\n",
    "\n",
    "# Mutational data\n",
    "sample_data = pd.DataFrame()\n",
    "sample_data[\"sample\"] = o_final_data[\"samplename_alias\"]\n",
    "sample_data[\"category\"] = o_final_data[\"Genes1\"]\n",
    "sample_data[\"value\"] = o_final_data[\"Variant Effect\"]\n",
    "sample_data[\"Histology\"] = o_final_data[\"Histology_\"]\n",
    "sample_data[\"grade\"] = o_final_data[\"GRAD\"]\n",
    "# sample_data[\"stage\"] = o_final_data[\"Stage_OC_I_IV\"]\n",
    "\n",
    "# sorting samples \n",
    "cancer_order = [\"HGSC\", \"LGSC\", \"Endometrioid\", \"Carcinosarcoma\", \"Mucinous\",\"Neuroendocrine carcinoma\", \"Carcinoma\", \"Teratome\", \"Unknown\"]\n",
    "\n",
    "df_mapping = pd.DataFrame({\n",
    "    'Histology': cancer_order,\n",
    "})\n",
    "sort_mapping = df_mapping.reset_index().set_index('Histology')\n",
    "\n",
    "sample_data[\"num\"] = sample_data[\"Histology\"].map(sort_mapping[\"index\"])\n",
    "sample_data[\"ID\"] = sample_data[\"sample\"].apply(lambda x: x[-2:]).astype(\"int64\")\n",
    "\n",
    "sample_data = sample_data.sort_values(by=\"num\")\n",
    "\n",
    "# Saving data\n",
    "# sample_data.to_excel(r\"sample_data.xlsx\")\n",
    "\n",
    "sample_data = pd.read_excel(r\"H:\\PhD\\Work_Packages\\Work_package1_Ascites\\Data\\sample_data_.xlsx\")\n",
    "sample_data[\"sample\"] = [\"T\"+ x[-2:] if x.startswith(\"T\") else \"A\"+x[-2:] for x in sample_data[\"sample\"]]\n",
    "\n",
    "# Using groupby & sort_values to sort.\n",
    "sample_data[\"ID\"] = sample_data.ID.astype(\"str\")\n",
    "\n",
    "final_data[\"sample_ID\"] = final_data.sample_ID.apply(lambda x: x.replace(\"_\", \"\")).astype(\"str\")\n",
    "sample_data = sample_data.loc[sample_data[\"ID\"].isin(final_data.sample_ID)]\n",
    "\n",
    "sample_data = sample_data.reindex(index =  [0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12, 13, 14, 15, 16,\n",
    "                                          17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33,\n",
    "                                          34, 35, 36, 37, 38, 39, 40, 41, 42, 43, 44, 45, 46, 47, 48, 49, 50,\n",
    "                                          51, 52, 53, 70, 71, 72, 73, 74, 75, 78, 79, 80, 81, 82, 83, 90, 91, 92, 84, 85,\n",
    "                                          86, 87])\n",
    "# Defining category order of genes\n",
    "category_order = ['CTNNB1','PTEN','FGFR2','MSH6','ATM','BRCA2','NF1','PIK3CA','KRAS','TP53']\n",
    "\n",
    "# Creating dictionary with colors and params\n",
    "mut_mapping = {'missense': '#00A893', \n",
    "               \"nonsense\": '#986B8C', \n",
    "               \"frameshiftDeletion\" : \"#F6931E\", \n",
    "               np.nan:\"grey\",\n",
    "               'unknown' : '#0b819c',\n",
    "               'Absent': {'facecolor': 'grey', 'alpha': 0.2}}\n",
    "\n",
    "###############################################################################"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4fcf5e72-8bfe-458e-9ab9-1601967c5855",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Creating indicator element\n",
    "\n",
    "indicator_df = pd.DataFrame()\n",
    "indicator_df[\"sample\"] = sample_data[\"sample\"].unique()\n",
    "indicator_df[\"group\"] = indicator_df[\"sample\"].apply(lambda x: x[-2:]).astype(\"int64\")\n",
    "indicator_df[\"category\"] = \"Same patient\"\n",
    "\n",
    "###############################################################################"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e6efa42c-c6d4-4586-8252-c3a4f403dcd7",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Creating Cancer type element\n",
    "\n",
    "cancer_type = sample_data[[\"sample\", \"Histology\"]]\n",
    "cancer_type = cancer_type.rename(columns = {\"Histology\":\"value\"})\n",
    "cancer_type[\"category\"] = \"Histology\"\n",
    "cancer_type = cancer_type.drop_duplicates(subset=[\"sample\"])\n",
    "\n",
    "###############################################################################"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1be6a4b2-a624-4399-8e9b-f2f4e635b035",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Creating side-bar chart element\n",
    "mut_freq = sample_data[\"category\"].value_counts()\n",
    "\n",
    "mut_freq = mut_freq.reset_index()\n",
    "mut_freq = mut_freq.rename(columns={\"index\":\"category\",\n",
    "                                    \"category\":\"Mutated samples\"})\n",
    "\n",
    "side_mapping = {'Mutated samples': 'darkgrey'}\n",
    "\n",
    "# calculate the percentage of samples with that gene mutated, rounding and adding a percent sign\n",
    "percentages = (mut_freq['Mutated samples']/86*100).round(1).astype(str) + '%'\n",
    "\n",
    "\n",
    "# decrease size of side bar\n",
    "widths = [0.4, 4.5]\n",
    "\n",
    "# move side bar plot slightly closer to comut\n",
    "wspace = 0.10\n",
    "\n",
    "###############################################################################"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "01cc825f-d73c-44dc-bcdb-41f5bb724719",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Creating grade element\n",
    "\n",
    "grade = sample_data[[\"sample\", \"grade\"]]\n",
    "grade = grade.drop_duplicates(subset=\"sample\")\n",
    "\n",
    "grading = pd.get_dummies(grade.grade)\n",
    "grading = grading.rename(columns={0:\"Unknown\",\n",
    "                                  1:\"Grade_1\",\n",
    "                                  2:\"Grade_2\",\n",
    "                                  3:\"Grade_3\",\n",
    "                                  4:\"Grade_4\"})\n",
    "\n",
    "grading[\"sample\"] = grade[\"sample\"]\n",
    "grading[\"Grade_4\"] = [4 if x == 1 else x for x in grading[\"Grade_4\"]]\n",
    "grading[\"Grade_3\"] = [3 if x == 1 else x for x in grading[\"Grade_3\"]]\n",
    "grading[\"Grade_2\"] = [2 if x == 1 else x for x in grading[\"Grade_2\"]]\n",
    "grading[\"Grade_1\"] = [1 if x == 1 else x for x in grading[\"Grade_1\"]]\n",
    "grading[\"Unknown\"] = [0 if x == 1 else x for x in grading[\"Unknown\"]]\n",
    "\n",
    "grading = grading[[\"sample\", \"Unknown\", \"Grade_1\", \"Grade_2\", \"Grade_3\", \"Grade_4\"]]\n",
    "\n",
    "bar_mapping = {\"Grade_4\": '#be2b9e', \"Grade_3\": '#7a3cae', \"Grade_2\" : '#dbdad9', \"Grade_1\": '#97a191', \"Unknown\":'white'}\n",
    "bar_kwargs = {'width': 0.8, 'edgecolor': 'black'}\n",
    "\n",
    "###############################################################################"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c726f839-6095-4b18-a3f0-db48cd0cbd8d",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Creating Stage element\n",
    "\n",
    "stage = pd.DataFrame()\n",
    "stage[\"sample\"] = o_final_data[\"samplename_alias\"]\n",
    "stage[\"sample\"] = [\"T\"+ x[-2:] if x.startswith(\"T\") else \"A\"+x[-2:] for x in stage[\"sample\"]]\n",
    "\n",
    "\n",
    "stage[\"stage\"] = o_final_data[\"Stage_OC_I_IV\"]\n",
    "stage[\"Stage_4\"] = [4 if x == \"IV\" else 0 for x in stage[\"stage\"]]\n",
    "stage[\"Stage_3\"] = [3 if x == \"III\" else 0 for x in stage[\"stage\"]]\n",
    "stage[\"Stage_2\"] = [2 if x == \"II\" else 0 for x in stage[\"stage\"]]\n",
    "stage[\"Stage_1\"] = [1 if x == \"I\" else 0 for x in stage[\"stage\"]]\n",
    "stage[\"Unknown\"] = [0 if x == np.nan else 0 for x in stage[\"stage\"]]\n",
    "\n",
    "stage = stage.drop_duplicates(subset=[\"sample\"])\n",
    "stage = stage.drop(columns=[\"stage\"])\n",
    "\n",
    "bar_mapping_stage = {\"Stage_4\": '#476930', \"Stage_3\": '#86B049', \"Stage_2\" : '#C8B88A', \"Stage_1\": '#F1DDDF', \"Unknown\":'white'}\n",
    "bar_kwargs = {'width': 0.8, 'edgecolor': 'black'}\n",
    "\n",
    "###############################################################################"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "40ac66b2-5a35-4030-bcd6-baa1ca7bf0d5",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Creating Overall survival element\n",
    "\n",
    "overall_survival = pd.DataFrame()\n",
    "overall_survival[\"sample\"] = o_final_data[\"samplename_alias\"]\n",
    "overall_survival[\"sample\"] = [\"T\"+ x[-2:] if x.startswith(\"T\") else \"A\"+x[-2:] for x in overall_survival[\"sample\"]]\n",
    "\n",
    "overall_survival[\"survival\"] = o_final_data[\"survival\"]\n",
    "overall_survival[\"survival\"].fillna(0, inplace=True)\n",
    "\n",
    "overall_survival = overall_survival.drop_duplicates(subset=[\"sample\"])\n",
    "bar_mapping_surv = {\"survival\": '#A62C2B'}\n",
    "bar_kwargs = {'width': 0.8, 'edgecolor': 'black'}\n",
    "\n",
    "###############################################################################"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3049d35c-bfdd-493b-90e3-66ea51fbf5b9",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Adjusting space between elements\n",
    "\n",
    "# decrease space between plots with space and wspace\n",
    "hspace = 0.08\n",
    "wspace = 0.1\n",
    "\n",
    "# increase height of biopsy site from default 1 and decrease burden height from default 3\n",
    "heights = {'Histology': 2, \"Same patient\" : 2}\n",
    "\n",
    "structure = [['Mutation type'],['Same patient'], [\"Histology\"], [\"Tumor Grade\"]]\n",
    "\n",
    "# decrease distance between purity and biopsy site\n",
    "subplot_hspace = 0.02\n",
    "\n",
    "###############################################################################"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4cf466a1-5c0e-48fb-ba75-104285cbdefa",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Plotting the co-mutational plot\n",
    "\n",
    "sns.set_theme(style=\"white\")\n",
    "\n",
    "ova_comut = comut.CoMut()\n",
    "\n",
    "# define order of comut BEFORE any data is added\n",
    "# ova_comut.samples = sample_order\n",
    "\n",
    "ova_comut.add_categorical_data(sample_data, name=\"Mutation type\", \n",
    "                             mapping=mut_mapping, \n",
    "                             category_order=category_order,\n",
    "                             tick_style=\"italic\")\n",
    "ova_comut.add_sample_indicators(indicator_df, name = \"Same patient\")\n",
    "ova_comut.add_categorical_data(cancer_type, name=\"Histology\", value_order=cancer_order)\n",
    "ova_comut.add_side_bar_data(mut_freq, paired_name = 'Mutation type', name = 'Mutated samples', position = 'left', \n",
    "                            mapping = side_mapping, xlabel = 'Mutated samples', bar_kwargs = {'alpha': 0.5})\n",
    "ova_comut.add_bar_data(grading, name = 'Tumor Grade', ylabel=\"Grade\", mapping=bar_mapping, stacked=True, bar_kwargs=bar_kwargs)\n",
    "ova_comut.add_bar_data(stage, name = 'Tumor Stage', ylabel=\"Stage\", mapping=bar_mapping_stage, stacked=True, bar_kwargs=bar_kwargs)\n",
    "ova_comut.add_bar_data(overall_survival, name = 'Survival', ylabel=\"Survival \\n (mth)\", mapping=bar_mapping_surv, stacked=True, bar_kwargs=bar_kwargs)\n",
    "\n",
    "\n",
    "ova_comut.plot_comut(figsize = (20, 6), x_padding = 0.04, y_padding = 0.04, tri_padding = 0.03, widths = widths, wspace = wspace, hspace=hspace)#, structure=structure)\n",
    "ova_comut.add_unified_legend(rename = {'Absent': 'Wild type'}, ncol=2)\n",
    "\n",
    "# Saving the plot\n",
    "#ova_comut.figure.savefig('ova_comut_updated.svg', bbox_inches = 'tight', dpi=600)\n",
    "\n",
    "###############################################################################"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
